Exception-Handling.md

```diff
@@ -0,0 +1,209 @@
+  puts "Made it past the error"
+end
+```
+The output would be:
+```bash
+$  ruby sample_exception.rb
+basic_exception_without_rescue.rb:4:in `/': divided by 0 (ZeroDivisionError)
+	from basic_exception_without_rescue.rb:4:in `<main>'
+```
+
+### Rescuing Specific Types of Exceptions
+
+There are a number of kinds of exceptions both for [regular'ol Ruby](https://ruby-doc.org/core-2.3.1/Exception.html) & [Rails](http://stackoverflow.com/a/19348733/918993).  Some types of errors we probably don't want to try to rescue because they are irrecoverable or so unexpected that we shouldn't write code to recover from them.  We can however set up specific blocks to recover from specific types of errors that we can and want to recover from.  Generally recoverable errors inherit from `StandardError` and for this reason when you place a rescue block without specifying the exception class Ruby will only handle exceptions inheriting from StandardError.  
+
+The syntax will look like this:
+
+```ruby
+begin
+  # some code that will create an exception.
+  
+rescue <some kind of exception>
+  # Code to recover from that exception
+
+rescue <some other kind of exception>
+  # Code to recover from another type of exception
+
+rescue
+  # other exceptions handled here
+
+ensure
+  #  This code runs no matter what.  
+
+end 
+```
+#### One Ruby Example - CSVReader
+
+In Ruby we could write code to handle problems, like in our CSV code to deal with problems, like being unable to open the CSV file.
+
+```ruby
+content = []
+begin
+  CSV.foreach("file.csv") do |row|
+
+    content << row
+  end
+
+  rescue SystemCallError
+    puts "I couldn't open the file."
+
+end
+```
+
+#### Exceptions Continue Up The Stack
+
+When an exception occurs, if it is not rescued in the immediate block or method it proceeds up and out of the method until the application crashes with an error message.
+
+```ruby
+def quotient_and_remainer dividend, divisor
+  answer = [dividend / divisor]
+  answer << quotient % divisor
+  answer
+end
+
+begin
+  puts quotient_and_remainer 10, 0
+rescue ZeroDivisionError
+  puts "Attempted to divide by zero"
+end
+```
+The method `divide` here will generate an exception when it tries to divide 10 by zero.  The exception is not "caught" in the method so ruby leaves the method and falls back to the caller of the method, which does have a rescue block to catch the `ZeroDivisionError`.  
+
+
+```bash
+$  ruby example.rb
+Attempted to divide by zero
+```
+
+
+#### Exception Handling in Rails
+
+In Rails we have a number of methods in ActiveRecord which can generate exceptions a few of which include:
+-  save!
+-  create!
+-  update!
+
+These methods trigger exceptions when validations fail or the database is unavailable.  So you could handle problems with creating a new task in TaskListRails using exceptions as follows.  
+
+```ruby
+  def create
+    @task = Task.new(task_params)
+
+    begin
+      @task.save!
+    rescue ActiveRecord::RecordInvalid => e
```

Comment:
Unless we've introduced the `=> e` syntax for getting access to the exception object earlier, I'm not sure it's a good idea to include it here.

We could probably set a generic flash message instead and have this be less confusing.